design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,276 @@
+    2. `stop_command`. This command carries out any required regardless of test outcome. It is important for the test runner to ensure that this is run on both success and failure
+    3. `service_type`. This is optional and for the future. The `service_type` specifies a internal implementation of a Service. The internal implementation come from a standard library provided by and maintened by the Infrastructure team.
+```
+{
+  "service_name": "mysql_service",
+  "service_type": "myql",
+  "start_command": "mysql_start.sh",
+  "stop_command": "mysql_stop.sh"
+}
+```
+
+A `Test Runner`  is software that provides a conventional way for defining and running test cases. They typically come with capabilities to measure test metrics, test outcomes and code coverage. This will also allow historical tracking. An example of a test-runner is Cucumber. Test cases can be defined as a sequence of steps in the Gherkin language. The step definitions are written in the language of the test runner e.g. Python, Ruby, Go. In this case we're strongly considering adopting `Cucumber` in Ruby as the test-runner. Ruby because of the ubiquity of working knowledge of that language in the team and the fact that there are Ruby Cucumber test suites that exist within some of our repos.
+
+The `Service Engine` is a combination of steps defined in the `Test Runner`. The Engine orchestrates the lifecycle of a Service. It provides step definitions for starting Services, stopping services regardless of the test outcome, creating named pipes for the Service output and writing that to the test context.
+
+A `Test Case` is a sequence of steps that goes through a scenario to validate some given functianlity. It should really include at least one step making an assertion as part of the validation.
+
+A `Context` is a temporary store within a test case. Steps can read and write from the store. The key `service_name` will be used to identify this output.
+
+Here is an example of the components in this architecture working together.
+
+1. Service definitions exist for Conjur and MySQL
+    1. In this example it is MySQL and Conjur. As mentioned in the description of Service Definition, output from `start_command` is written into the test case `context` under a path corresponding to `service_name`. An example for MySQL is provided below.
+    ```
+    {
+      "service_name": "mysql_service",
+      "service_type": "myql",
+      "start_command": "mysql_start.sh", # imagine that "mysql_start.sh" calls out to "helm install"
+      "stop_command": "mysql_stop.sh"
+    }
+    ```
+1. Service Configs exist for Conjur and MySQL. These files would exist as `./mysql_config.json` and `./conjur_config.json`. An example for Conjur is provided below
```

Comment:
I don't know if I like completely unique service definitions - is there stuff we can reuse?